Objects/interpolationobject.c

```diff
@@ -0,0 +1,212 @@
+{
+    interpolationobject *self = interpolationobject_CAST(op);
+    return PyUnicode_FromFormat("%s(%R, %R, %R, %R)",
+                                _PyType_Name(Py_TYPE(self)), self->value, self->expression,
+                                self->conversion, self->format_spec);
+}
+
+static PyMemberDef interpolation_members[] = {
+    {"value", Py_T_OBJECT_EX, offsetof(interpolationobject, value), Py_READONLY, "Value"},
+    {"expression", Py_T_OBJECT_EX, offsetof(interpolationobject, expression), Py_READONLY, "Expression"},
+    {"conversion", Py_T_OBJECT_EX, offsetof(interpolationobject, conversion), Py_READONLY, "Conversion"},
+    {"format_spec", Py_T_OBJECT_EX, offsetof(interpolationobject, format_spec), Py_READONLY, "Format specifier"},
+    {NULL}
+};
+
+PyTypeObject _PyInterpolation_Type = {
+    PyVarObject_HEAD_INIT(NULL, 0)
+    .tp_name = "string.templatelib.Interpolation",
+    .tp_doc = PyDoc_STR("Interpolation object"),
+    .tp_basicsize = sizeof(interpolationobject),
+    .tp_itemsize = 0,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+    .tp_new = interpolation_new,
+    .tp_alloc = PyType_GenericAlloc,
+    .tp_dealloc = interpolation_dealloc,
+    .tp_clear = interpolation_clear,
+    .tp_free = PyObject_GC_Del,
+    .tp_repr = interpolation_repr,
+    .tp_members = interpolation_members,
+    .tp_traverse = interpolation_traverse,
+};
+
+PyStatus
+_PyInterpolation_InitTypes(PyInterpreterState *interp)
+{
+    PyObject *tuple = Py_BuildValue("(ssss)", "value", "expression", "conversion", "format_spec");
+    if (!tuple) {
+        goto error;
+    }
+
+    PyObject *dict = _PyType_GetDict(&_PyInterpolation_Type);
+    if (!dict) {
+        Py_DECREF(tuple);
+        goto error;
+    }
+
+    int status = PyDict_SetItemString(dict, "__match_args__", tuple);
+    Py_DECREF(tuple);
+    if (status < 0) {
+        goto error;
+    }
+    return _PyStatus_OK();
+
+error:
+    return _PyStatus_ERR("Can't initialize interpolation types");
+}
+
+PyObject *
+_PyInterpolation_Build(PyObject *value, PyObject *str, int conversion, PyObject *format_spec)
+{
+    interpolationobject *interpolation = PyObject_GC_New(interpolationobject, &_PyInterpolation_Type);
+    if (!interpolation) {
+        return NULL;
+    }
+
+    interpolation->value = Py_NewRef(value);
+    interpolation->expression = Py_NewRef(str);
+    interpolation->format_spec = Py_NewRef(format_spec);
+
+    if (conversion == 0) {
+        interpolation->conversion = Py_None;
+    }
+    else {
+        switch (conversion) {
+            case FVC_ASCII:
+                interpolation->conversion = _Py_LATIN1_CHR('a');
+                break;
+            case FVC_REPR:
+                interpolation->conversion = _Py_LATIN1_CHR('r');
+                break;
+            case FVC_STR:
+                interpolation->conversion = _Py_LATIN1_CHR('s');
+                break;
+            default:
+                PyErr_SetString(PyExc_SystemError,
```

Comment:
We need to set `conversion` to `NULL` here. `PyObject_GC_New` does not zero-out the allocation, so `conversion` is just garbage memory right now. Trying to `Py_DECREF` that (via the interpolation's finalizer) will trigger a segfault.
```suggestion
                interpolation->conversion = NULL;
                PyErr_SetString(PyExc_SystemError,
```

Reply:
That's right! Done.